### PR TITLE
REDCap Connection

### PR DESCRIPTION
**Pull request by aditya10**
General support for REDCap EM:
- [x] New selector system
- [x] connect REDCap and parse selectors
- [x] Implement federated sum, count using optimized SQL statements

SQL Generators and API:
- [x] Site administrators can define custom SQL statement generators for REDCap
- [x] A new API to verify whether the `sql-options` are valid 
- [ ] Add string escaping and validation at the coordinator level
- [ ] Bug fixes

### CODE REVIEW

**Comment by bestchai** [comment]
Should we merge this PR so that it's not outstanding? Or would you prefer to keep adding commits to the PR over the course of the summer? :-)

**Comment by aditya10** [comment]
> Should we merge this PR so that it's not outstanding? Or would you prefer to keep adding commits to the PR over the course of the summer? :-)

After I finish the part on performance testing, I want to spend some time updating this code and integrating the new changes I made in the last week or so. Then I'll get it reviewed and merged :)

File: cloudalgo/cloud_algo.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import logging
 from pylogrus import PyLogrus, TextFormatter
 import utils.env_manager as env_manager
+import random
```
Comment (stolet) [inline cloudalgo/cloud_algo.py:11]: I think this import is unused.

File: api/leap.py [hunk 2]
```diff
@@ -133,3 +133,55 @@ def _get_compute_stub(self):
     def _create_request_obj(self):
         request = pb.computation_msgs_pb2.ComputeRequest()
         return request
+
```
Comment (stolet) [inline api/leap.py:136]: Inherit from the DistributedLeap class. Some of the attributes in the class are the same and the get_compute_stub function is the same, so you can reuse that. The others you can override as you wish.

File: sitealgo/site_algo.py [hunk 3]
```diff
@@ -180,48 +233,113 @@ def map_logic(self, request):
...
+        if self.config["csv_true"] == "1" or selector["type"] == "csv":
+            return self.get_csv_data(selector)
         else:
-            return self.get_redcap_data(redCapUrl, redCapToken, filter)
+            return self.get_redcap_data(redCapUrl, redCapToken, selector)
 
 
     # TODO: Actually filter the data according to a user selector
     # Gets the data from a csv file and returns the records to
     # perform a computation on.
```
Comment (stolet) [inline sitealgo/site_algo.py:276]: Add comment describing the parameters of the function.

File: sitealgo/site_algo.py [hunk 4]
```diff
@@ -180,48 +233,113 @@ def map_logic(self, request):
...
     #
     # url: Url of the RedCap project
     # token: Token used to access RedCap project given in the url
-    # filterLogic: The filter to be applied to the results."""
-    def get_redcap_data(self, url, token, filter_logic):
-        # project = redcap.Project(url, token)
-        # patients = project.export_records(filter_logic=filter_logic)
-        # return patients
-        return [1,2,3,4]
+    # selector:
```
Comment (stolet) [inline sitealgo/site_algo.py:292]: Add comment describing the parameters of the function.

File: sitealgo/site_algo.py [hunk 5]
```diff
@@ -180,48 +233,113 @@ def map_logic(self, request):
...
+        elif selector["type"] == "sql":
+            query_gen_func = rc_sql_gen.generator_map[selector["sql_func"]](selector["sql_options"])
+            query = query_gen_func["generate"]()
+            return self.get_redcap_query_result(query)
+        return None
+
+    # Contacts a redCap project and returns the filtered records
+    # from this project
+    #
+    # filterLogic: The filter to be applied to the results.
```
Comment (stolet) [inline sitealgo/site_algo.py:305]: Add comment describing selected_fields parameter.

File: api/leap.py [hunk 6]
```diff
@@ -133,3 +133,55 @@ def _get_compute_stub(self):
...
+class DistributedSelectorVerification():
```
Comment (stolet) [inline api/leap.py:137]: Add comment describing the parameters of each function in this class.

**Review by stolet — APPROVED** [review]
LGTM

**Pull request merged**